mbq/atomiq/consumers.py

```diff
@@ -14,22 +14,12 @@
 
 class BaseConsumer(object):
 
-    def run(self):
-        try:
-            with transaction.atomic():
-                task = self.model.objects.available_for_processing()[:1].select_for_update().get()
-                execution_started_at = arrow.utcnow()
-                self.process_task(task)
-                execution_ended_at = arrow.utcnow()
-            self.send_task_execution_metrics(task, execution_started_at, execution_ended_at)
-        except self.model.DoesNotExist:
-            sleep(0.5)
-        except Exception:
-            rollbar.report_exc_info()
-            sleep(0.5)
-
-    def process_task(self, task):
+    @transaction.atomic
+    def process_one_task(self):
+        task = self.model.objects.available_for_processing()[:1].select_for_update().get()
```

Comment:
Seems weird for this function to very frequently throw a not found exception, maybe `return None` or something instead? Then the consumer still knows it needs to sleep but without having to do this function's error handling for it

Reply:
why is that a concern? It makes the locking and checking for the task a single db operation. to return none it would mean I would just need to wrap it in the same exception handling just on the consumer side.

Reply:
Checking for None isn't exception handling, `if processed_task:` is a lot more natural than a try-except with a specific error class

Ultimately it works either way but not being able to call a function unless you know its specific error handling expectations is much less user-friendly than a function that returns a value or None, and only throws if something really went wrong

Reply:
> I mean even for loops are implemented using exceptions.

....what? I didn't follow this at all

> One thing I will say though is it could be good to raise a more descriptive error, like "NoAvailableTasksToProcess" so the intent of the code is more clearly expressed through the exception.

+1, throwing an exception from our lib is much better than bubbling up an exception from a library we're using. Better to not expose implementation details like that. I think either returning None or a custom exception works well here.